sdk/typescript/src/multisig/publickey.ts

```diff
@@ -0,0 +1,260 @@
+		super();
+
+		if (typeof value === 'string') {
+			this.rawBytes = fromB64(value);
+			this.multisigPublicKey = builder.de('MultiSigPublicKey', this.rawBytes);
+		} else if (value instanceof Uint8Array) {
+			this.rawBytes = value;
+			this.multisigPublicKey = builder.de('MultiSigPublicKey', this.rawBytes);
+		} else {
+			this.multisigPublicKey = value;
+			this.rawBytes = builder.ser('MultiSigPublicKey', value).toBytes();
+		}
+
+		this.publicKeys = this.multisigPublicKey.pk_map.map(({ pubKey, weight }) => {
+			const [scheme, bytes] = Object.entries(pubKey)[0] as [SignatureScheme, number[]];
+			return {
+				publicKey: publicKeyFromBytes(scheme, Uint8Array.from(bytes)),
+				weight,
+			};
+		});
+
+		if (this.publicKeys.length > MAX_SIGNER_IN_MULTISIG) {
+			throw new Error(`Max number of signers in a multisig is ${MAX_SIGNER_IN_MULTISIG}`);
+		}
+	}
+
+	static fromPublicKeys(threshold: number, publicKeys: { publicKey: PublicKey; weight: number }[]) {
+		return new MultiSigPublicKey({
+			pk_map: publicKeys.map(({ publicKey, weight }) => {
+				const scheme = SIGNATURE_FLAG_TO_SCHEME[publicKey.flag() as SignatureFlag];
+
+				return {
+					pubKey: { [scheme]: Array.from(publicKey.toBytes()) } as PublicKeyEnum,
+					weight,
+				};
+			}),
+			threshold,
+		});
+	}
+
+	/**
+	 * Checks if two MultiSig public keys are equal
+	 */
+	override equals(publicKey: MultiSigPublicKey): boolean {
+		return super.equals(publicKey);
+	}
+
+	/**
+	 * Return the byte array representation of the MultiSig public key
+	 */
+	toBytes(): Uint8Array {
+		return this.rawBytes;
+	}
+
+	/**
+	 * Return the Sui address associated with this MultiSig public key
+	 */
+	toSuiAddress(): string {
+		// max length = 1 flag byte + (max pk size + max weight size (u8)) * max signer size + 2 threshold bytes (u16)
+		const maxLength = 1 + (64 + 1) * MAX_SIGNER_IN_MULTISIG + 2;
+		const tmp = new Uint8Array(maxLength);
+		tmp.set([SIGNATURE_SCHEME_TO_FLAG['MultiSig']]);
+
+		tmp.set(bcs.ser('u16', this.multisigPublicKey.threshold).toBytes(), 1);
+		let i = 3;
+		for (const { publicKey, weight } of this.publicKeys) {
+			tmp.set([publicKey.flag()], i);
+			tmp.set(publicKey.toBytes(), i + 1);
+			tmp.set([weight], i + 1 + publicKey.toBytes().length);
+			i += publicKey.toBytes().length + 2;
+		}
+		return normalizeSuiAddress(bytesToHex(blake2b(tmp.slice(0, i), { dkLen: 32 })));
+	}
+
+	/**
+	 * Return the Sui address associated with this MultiSig public key
+	 */
+	flag(): number {
+		return SIGNATURE_SCHEME_TO_FLAG['MultiSig'];
+	}
+
+	/**
+	 * Verifies that the signature is valid for for the provided message
+	 */
+	async verify(message: Uint8Array, multisigSignature: Uint8Array): Promise<boolean> {
+		const multisig = builder.de('MultiSig', multisigSignature);
+		let signatureWeight = 0;
+
+		if (
+			!bytesEqual(
+				builder.ser('MultiSigPublicKey', this.multisigPublicKey).toBytes(),
+				builder.ser('MultiSigPublicKey', multisig.multisig_pk).toBytes(),
```

Comment:
the whole verify method is a bit of a hack, and might be an indicator that this is the wrong abstraction.  The `multisigSignature` is the full serializedSignature excluding the flag (this includes the publicKey portion).  For other keypairs, we verify against the signature portion only, but there is no single signature we can pass here, since we decode directly into the MultiSig bcs struct which has a publicKey, threshold, and sigs[].

This check is the easiest way to check the public keys for the provided signature, and the current publicKey match, since `parsePartialSignatures` will extract the publicKeys from the provided signature rather than the current publicKey.  We could instead move parsePartialSignatures onto the class and have it read publicKeys and weights from the class rather than the provided signature